File: python/moneysocket/utl/bech32.py
# Copyright (c) 2020 Jarret Dyrbye
# Distributed under the MIT software license, see the accompanying
# file LICENSE or http://www.opensource.org/licenses/mit-license.php

import bitstring
from bech32 import bech32_encode, bech32_decode, convertbits


class Bech32():
    @staticmethod
    def encode_bytes(bytes_to_encode, hrp):
        converted = convertbits(bytes_to_encode, 8, 5, True)
        assert converted
        return bech32_encode(hrp, converted)

    @staticmethod
    def decode_bytes(bech32_string):
        try:
            hrp, data = bech32_decode(bech32_string)
        except:
            return None, None
        if not hrp:
            return None, None
        deconverted = convertbits(data, 5, 8, False)
        assert deconverted
        decoded = bytes(deconverted)
        return hrp, decoded

File: browser/src/js/moneysocket/utl/bech32.js
// Copyright (c) 2020 Jarret Dyrbye
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php

const bech32 = require('bech32');

const BECH32_LIMIT = 20000;

class Bech32 {
    static encodeBytes(bytes, hrp) {
        var words = bech32.toWords(bytes);
        return bech32.encode(hrp, words, BECH32_LIMIT);
    }

    static decodeBytes(bech32_str) {
        var d;
        try {
            d = bech32.decode(bech32_str, BECH32_LIMIT);
        } catch(err) {
            return [null, null];
        }

        var words = d.words;
        var hrp = d.prefix;

        var bytes;
        try {
            bytes = bech32.fromWords(words);
        } catch(err) {
            return [null, null];
        }
        return [hrp, bytes];
    }
}

exports.Bech32 = Bech32;
